Util/src/main/java/io/deephaven/util/codec/ObjectDecoder.java

```diff
@@ -30,12 +32,30 @@ public interface ObjectDecoder<TYPE> {
      *
      * @param input The input byte array containing bytes to decode
      * @param offset The offset into the byte array to start decoding from
-     * @param length The length of the byte array to decode from, starting at the offset
+     * @param length The number of bytes to decode, starting at the offset
      * @return The output object, possibly null
      */
     @Nullable
     TYPE decode(@NotNull byte[] input, int offset, int length);
 
+    /**
+     * Decode an object from a ByteBuffer.
+     *
+     * @param buffer The input ByteBuffer containing bytes to decode
+     * @return The output object, possibly null
+     */
+    @Nullable
+    default TYPE decode(@NotNull final ByteBuffer buffer) {
+        if (buffer.hasArray()) {
+            return decode(buffer.array(), buffer.arrayOffset() + buffer.position(), buffer.remaining());
+        } else {
+            // Make a copy of the buffer's contents
+            final byte[] bytes = new byte[buffer.remaining()];
+            buffer.get(bytes);
+            return decode(bytes, 0, bytes.length);
+        }
```

Comment:
There's a slight difference in semantics between these two cases; in the array case, we aren't modifying the position, and in the copy case we are. I think we should either:

* Update position in both cases and document behavior
* Don't update position in both cases and document behavior
* Document behavior as "may or may not" modify the buffer position

@rcaudy preferences?

Reply:
Good question. Any of the above is "reasonable", but 2 is the least "normal".

Reply:
I think we should pick 3. Put the onus on the user if they care.